### ISSUE TITLE
Bad (1200ms) 'add appointment' UI performance

### ISSUE THREAD

**Issue by jasperdj** [KinToday/kin-web-client#42]
I'd like to share my elaborate feedback on your UI performance. 

The UI is excellent and I'm a big fan. I want to use this product for a very long time. However, there is one big UX bottleneck. Adding appointments is slow and clunky. By that, I mean the delay between clicking on the calendar to add an appointment and the popup actually showing up. Here's my analysis.

The total duration between mouse-down and popup showing is 1200 milliseconds. This is too long to show a simple popup. It especially becomes a bottleneck when one is clicking up and down to set the right appointment frame. It feels clunky and poorly engineered. 

What's the cause? Well, 776.3 ms is used on rendering. But why?

Finding the root cause was difficult to find since all the sources are minimized into one file on the public server. And I don't have time enough to go through the source code. Based on a quick glance on the call tree I can make a few assertions. 
- 465 ms is used on your '_reducer_add_events' state reducer. See code section 1.
- 562 ms is used on a middleware call. See code section 2.

Conclusion, the root cause is likely to be due to inefficient and/or blocking state management. 

IF you don't have enough human resources to fix this within a few months I'm willing to contribute - let me know! In any case, I care about this product and would love to see it evolve. 

Thank you, 
Jasper de Jong 

Code section 1
 ```
key: "_reducer_add_events",
                    value: function(e, t) {
                        var o = this;
                        if (t.creating) {
                            var r = f.a.map(t.events, f.a.partial(a, m.c.timezone))
                              , l = i({
                                id: "kin-0:meta-create-layer",
                                acl: {
                                    create: !1,
                                    edit: !1,
                                    delete: !1
                                },
                                color: "#EC4956",
                                text_color: "#FFFFFF",
                                events: [],
                                selected: !0,
                                loaded: !0
                            });
```

Code section 2
```
key: "middleware",
                    value: function(e) {
                        var t = this;
                        return function(n) {
                            return function(o) {
                                var r = e.getState()
                                  , i = f.a.get(t._reducer, o.type);
                                f.a.isFunction(i) && i.call(t, r, o),
                                n(o)
                            }
                        }
                    }
```

**Comment by zowesiouff** [comment]
> Conclusion, the root cause is likely to be due to inefficient and/or blocking state management.

Yes the main issue is the rendering layer that we use ( FullCalendar ) which, last time I checked, didn't really support incremental updates to a single event. The entry point is the "middleware" starting here: [calendar.js](https://github.com/KinToday/kin-web-client/blob/03f95e6892b641e6fc280440942f75ae0a36dcc2/src/client/calendar.js#L219) we can most likely have huge gains there but it's at the expense of more complexity / or a better rendering layer which are both extremely time consuming :/ 

> IF you don't have enough human resources to fix this within a few months I'm willing to contribute - let me know! In any case, I care about this product and would love to see it evolve.

I would love to, hit me up with more questions if you need any help, I should be more responsive in the next couple of months ;)
